### PR TITLE
160 clean up abstractspotanalysisimageprocessor

### PR DESCRIPTION
**Pull request by bbean23**
This PR is a biggie. It includes:

* combining AbstractSpotAnalysisImageProcessor and it's ...Leger class
* moving much of the cacheing logic to CacheableImage
* adding lots of tests
* code cleanup and documentation

### CODE REVIEW

**Comment by rcbrost** [comment]
Randy Brost started reviewing this on Thursday October 31 at 9:27 AM.

File: opencsp/common/lib/cv/test/test_CacheableImage.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+        self.assertEqual(c2, CacheableImage.lru(True))
+        # LRU: 3
+        c3.nparray
+        self.assertEqual(c3, CacheableImage.lru(True))
+
+        # check that there are no more registered cacheable images
+        self.assertEqual(None, CacheableImage.lru(False))
+
+    def test_save_image(self):
+        """Verify that after saving the image, cacheing no longer creates a cache file"""
```
Comment (e10harvey) [inline opencsp/common/lib/cv/test/test_CacheableImage.py:275]: Should this (and can it) check whether the in-memory np array & image are 0 bytes?

**Review by e10harvey — CHANGES_REQUESTED** [review]
Thank you for these changes, @bbean23. Most of my comments are related to the end-user documentation. However, I think given the nature of these changes, it makes sense to share before and after performance metrics. Is there a example or test you could run to share memory usage and runtime before & after these changes?

File: opencsp/common/lib/cv/CacheableImage.py [hunk 2]
```diff
@@ -92,58 +395,202 @@ def _load_image(im: str | np.ndarray) -> npt.NDArray[np.int_]:
...
+        self._source_path = image_path_name_ext
+
+    @staticmethod
+    def cache_images_to_disk_as_necessary(
+        memory_limit_bytes: int, tmp_path_generator: Callable[[], str], log_level=lt.log.DEBUG
+    ):
+        """Check memory usage and convert images to files (aka file path
+        strings) as necessary in order to reduce memory usage."""
+        total_mem_size = CacheableImage.all_cacheable_images_size()
+        if total_mem_size <= memory_limit_bytes:
```
[the remark this answers is not in the document]
Comment (bbean23) [reply opencsp/common/lib/cv/CacheableImage.py:623]: I disagree. By providing the `memory_limit_bytes` as a parameter, we're effectively enabling the user to choose a lower memory threshold than is the default. There's also the benefit of requiring the user to think about how much memory they want to use, which is going to be system and application specific.

[the remark this answers is not in the document]
Comment (bbean23) [reply opencsp/common/lib/cv/CacheableImage.py:623]: added this same comment to the code

File: opencsp/common/lib/cv/CacheableImage.py [hunk 3]
```diff
@@ -92,58 +395,202 @@ def _load_image(im: str | np.ndarray) -> npt.NDArray[np.int_]:
...
+        """
+        self.to_image().save(image_path_name_ext)
+        self._source_path = image_path_name_ext
+
+    @staticmethod
+    def cache_images_to_disk_as_necessary(
+        memory_limit_bytes: int, tmp_path_generator: Callable[[], str], log_level=lt.log.DEBUG
+    ):
+        """Check memory usage and convert images to files (aka file path
+        strings) as necessary in order to reduce memory usage."""
```
[the remark this answers is not in the document]
Comment (bbean23) [reply opencsp/common/lib/cv/CacheableImage.py]: done

File: opencsp/common/lib/cv/CacheableImage.py [hunk 4]
```diff
@@ -92,58 +395,202 @@ def _load_image(im: str | np.ndarray) -> npt.NDArray[np.int_]:
...
         self._array = None
         self._image = None
-        self.cached = True
+
+    def save_image(self, image_path_name_ext: str):
+        """
+        Saves this image as an image file to the given file.
+
+        Note: this replaces the internal reference to source_path, if any, with
+        the newly given path.
```
[the remark this answers is not in the document]
Comment (bbean23) [reply opencsp/common/lib/cv/CacheableImage.py]: Actually, going back in and expanding the documentation on this saved me from a potential bug! Thanks!

File: opencsp/common/lib/cv/CacheableImage.py [hunk 5]
```diff
@@ -92,58 +395,202 @@ def _load_image(im: str | np.ndarray) -> npt.NDArray[np.int_]:
...
+                        + "Cacheing requires at least one path (cache_path or source_path) to be non-None!",
+                    )
+
+        # Cache this instance.
+        if self._does_source_image_match(self.nparray):
+            # This instance was created from an image file, so we can simply
+            # depend on that image file instead of writing a new numpy file to
+            # disk.
+            pass
+        elif ft.file_exists(self.cache_path) and os.path.getsize(self.cache_path) > 10:
```
[the remark this answers is not in the document]
Comment (bbean23) [reply opencsp/common/lib/cv/CacheableImage.py:546]: done! (the answer is paranoia)

File: opencsp/common/lib/cv/CacheableImage.py [hunk 6]
```diff
@@ -92,58 +395,202 @@ def _load_image(im: str | np.ndarray) -> npt.NDArray[np.int_]:
...
+        """Returns true if this image's source_path image file matches the data
+        in the given numpy array."""
+        if self._source_path is not None and ft.file_exists(self._source_path):
+            imarray = np.array(Image.open(self._source_path))
+            try:
+                arrays_are_equal = np.equal(nparray, imarray).all()
+            except Exception:
+                return False
+
+            enforce_source_image_matches = True
```
[the remark this answers is not in the document]
Comment (bbean23) [reply opencsp/common/lib/cv/CacheableImage.py]: done

**Review by bbean23 — COMMENTED** [review]
I believe I've addressed all your comments. Please approve when ready.

**Review by e10harvey — APPROVED** [review]

**Pull request merged**